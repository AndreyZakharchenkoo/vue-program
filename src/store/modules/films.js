import ApiService from '@/core/api';
import { SEARCH_PARAMS, SORT_PARAMS } from '@/core/constants';

export default {
  state: {
    searchValue: '',
    searchParam: SEARCH_PARAMS.title,
    sortParam: SORT_PARAMS.date,
    currentFilm: {},
    films: [],
    totalFoundedFilms: 0,
    offset: 0,
  },
  getters: {},
  mutations: {
    SET_FILMS(state, payload) {
      state.films = payload;
    },
    SET_TOTAL_FOUNDED_FILMS(state, payload) {
      state.totalFoundedFilms = payload;
    },
    SET_CURRENT_FILM(state, payload) {
      state.currentFilm = payload;
    },
    UPDATE_SEARCH_VALUE(state, payload) {
      state.searchValue = payload;
    },
    UPDATE_SEARCH_PARAM(state, payload) {
      state.searchParam = payload;
    },
    UPDATE_SORT_PARAM(state, payload) {
      state.sortParam = payload;
    },
    UPDATE_OFFSET(state, payload) {
      state.offset = payload;
    },
  },
  actions: {
    FETCH_FILMS_BY_PARAMS({ commit }, params) {
      ApiService.fetchFilmsByParams(params)
        .then((res) => {
          commit('SET_FILMS', res.data.data);
          commit('SET_TOTAL_FOUNDED_FILMS', res.data.total);
        })
        .catch((err) => {
          throw err;
        });
    },
    FETCH_FILM_BY_ID({ commit }, id) {
      ApiService.fetchFilmById(id)
        .then((res) => {
          commit('SET_CURRENT_FILM', res.data);
        })
        .catch((err) => {
          throw err;
        });
    },
  },
};
