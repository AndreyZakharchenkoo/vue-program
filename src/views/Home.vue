<template>
  <div>
    <section class="box-search">
      <div class="container">
        <h1>FIND YOUR MOVIE</h1>
        <form
          class="search"
          @submit.prevent>
          <div class="search__input">
            <the-input placeholder="Search" v-model.trim="searchValue" />
          </div>
          <the-button type="submit" class="large">Search</the-button>
        </form>
        <div class="sort">
          <div class="sort__label">Search by</div>
          <div class="button-group">
            <the-button class="active" @click="searchByParam('title', $event)">Title</the-button>
            <the-button @click="searchByParam('genres', $event)">Genre</the-button>
          </div>
        </div>
      </div>
    </section>
    <section class="box-gallery">
      <div class="box-gallery__top">
        <div class="container">
          <div v-if="films.length && searchValue" class="quantity">{{ films.length }} movie found</div>
          <div class="sort">
            <div class="sort__label">Sort by</div>
            <div class="button-group">
              <the-button class="active" @click="sortByParam('date', $event)">Release date</the-button>
              <the-button @click="sortByParam('rating', $event)">Rating</the-button>
            </div>
          </div>
        </div>
      </div>
      <div class="container">
        <ul v-if="films.length" class="gallery">
          <li
            class="post__item"
            v-for="item in films"
            :key="item.id"
          >
            <the-card :card="item" />
          </li>
        </ul>
        <the-placeholder
          v-else
          title="No films found"
        />
      </div>
    </section>
  </div>
</template>

<script>
import { mapGetters, mapMutations } from 'vuex'
import TheInput from '@/stories/TheInput.vue'
import TheButton from '@/stories/TheButton.vue'
import TheCard from '@/stories/TheCard.vue'
import ThePlaceholder from '@/stories/ThePlaceholder.vue'

export default {
  name: 'Home',
  components: {
    TheButton,
    TheInput,
    TheCard,
    ThePlaceholder
  },
  data: () => ({
    searchType: 'title',
    sortParam: 'date'
  }),
  computed: {
    ...mapGetters({
      films: 'GET_FILMS'
    }),
    searchValue: {
      get () {
        return this.$store.state.searchValue
      },
      set (val) {
        this.UPDATE_SEARCH_VALUE(val)
      }
    }
  },
  methods: {
    ...mapMutations([
      'UPDATE_SEARCH_VALUE',
      'UPDATE_SEARCH_PARAM',
      'UPDATE_SORT_PARAM'
    ]),
    switchActiveButton (event) {
      const buttonsGroup = event.path[1].querySelectorAll('button')
      buttonsGroup.forEach((elem) => {
        elem.classList.remove('active')
      })
      event.target.classList.add('active')
    },
    searchByParam (param, event) {
      this.switchActiveButton(event)
      this.UPDATE_SEARCH_PARAM(param)
    },
    sortByParam (param, event) {
      this.switchActiveButton(event)
      this.UPDATE_SORT_PARAM(param)
    }
  }
}
</script>

<style lang="scss" scoped>
.box-search {
  .search {
    margin: 40px 0 25px;
  }
}

.search {
  display: grid;
  grid-gap: 15px;
  grid-template-columns: 1fr min-content;
}

.quantity {
  font-weight: bold;
}

.sort {
  display: flex;
  align-items: center;

  &:only-child {
    margin-left: auto;
  }

  &__label {
    margin-right: 16px;
    opacity: 0.8;
    font-size: 16px;
    font-weight: 500;
    letter-spacing: 0.89px;
    line-height: 19px;
    text-transform: uppercase;
  }
}
</style>
