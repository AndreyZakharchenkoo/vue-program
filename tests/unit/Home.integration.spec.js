import { createLocalVue, mount } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import films from '@/store/modules/films';
// import filmsData from './mock-data/filmsData.json';
import lazyLoadImg from '@/plugins/lazy-loading';
import Home from '@/pages/home.vue';

global.fetch = require('jest-fetch-mock');

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(lazyLoadImg);
localVue.use(Vuex);

describe('Home.vue', () => {
  let wrapper;
  let store;

  beforeEach(() => {
    // router
    const router = new VueRouter({
      routes: [
        {
          path: '/',
          name: 'home',
          component: Home,
        },
      ],
    });
    // router.push('http://localhost:8080/?limit=9&searchVal=&searchBy=title&sortBy=release_date&offset=0');

    // store
    store = new Vuex.Store({
      modules: {
        films,
      },
    });

    wrapper = mount(Home, {
      localVue,
      store,
      router,
      stubs: ['router-link', 'router-view'],
    });

    fetch.resetMocks();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Renders correctly', () => {
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  it('Action "FETCH_FILMS_BY_PARAMS" executes after component was mounted', async () => {
    // fetch.once(JSON.stringify(filmsData));
    const searchInput = wrapper.find('.search__input input');
    expect(searchInput.exists()).toBe(true);
    // await searchInput.setValue('fif');

    // await localVue.nextTick();
    // await localVue.nextTick();

    // const filmsList = wrapper.find('.gallery');
    //
    // expect(fetch).toBeCalledWith('http://react-cdp-api.herokuapp.com/movies?limit=9&searchBy=title&search=fif&sortBy=release_date&offset=0&sortOrder=asc');
    // expect(filmsList.exists()).toBe(true);
  });
});
