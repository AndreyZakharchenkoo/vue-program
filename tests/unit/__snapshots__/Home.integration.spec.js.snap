// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue Renders correctly 1`] = `
<div>
  <section
    class="box-search"
  >
    <div
      class="container"
    >
      <h1>
        FIND YOUR MOVIE
      </h1>
       
      <form
        class="search"
      >
        <div
          class="search__input"
        >
          <input
            class="input"
            placeholder="Search"
            type="text"
          />
        </div>
         
        <button
          class="button large"
          type="submit"
        >
          Search
        </button>
      </form>
       
      <div
        class="sort"
      >
        <div
          class="sort__label"
        >
          Search by
        </div>
         
        <div
          class="button-group"
        >
          <button
            class="button active"
            type="button"
          >
            
            Title
          
          </button>
           
          <button
            class="button"
            type="button"
          >
            
            Genre
          
          </button>
        </div>
      </div>
    </div>
  </section>
   
  <section
    class="box-gallery"
  >
    <div
      class="box-gallery__top"
    >
      <div
        class="container"
      >
        <!---->
         
        <div
          class="sort"
        >
          <div
            class="sort__label"
          >
            Sort by
          </div>
           
          <div
            class="button-group"
          >
            <button
              class="button active"
              type="button"
            >
              
              Release date
            
            </button>
             
            <button
              class="button"
              type="button"
            >
              
              Rating
            
            </button>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="container"
    >
      <div
        class="placeholder"
      >
        <h2>
          No films found
        </h2>
      </div>
    </div>
  </section>
</div>
`;
